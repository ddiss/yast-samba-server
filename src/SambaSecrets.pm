#! /usr/bin/perl
# File:		modules/SambaSecrets.pm
# Package:	Samba server
# Summary:	Reading of /etc/samba/secrets.tdb
# Authors:	Stanislav Visnovsky <visnov@suse.cz>
#
# $Id$
#

package SambaSecrets;

use strict;
use Data::Dumper;

use YaST::YCP qw(:DATA :LOGGING);
use YaPI;

textdomain "samba-server";
our %TYPEINFO;

## Global imports
BEGIN{
YaST::YCP::Import("SCR");
}

use constant {
    secret_tdb => "/etc/samba/secrets.tdb",
    tdbdump => "/usr/bin/tdbdump",
};

my $Secrets;

# return true if effective user id is 0 (root user)
sub isRoot {
    return 1; #TODO
}

# Read the current contents of the secrets file
BEGIN{$TYPEINFO{Read}=["function", "boolean"]}
sub Read {
    my ($self) = shift;
    
    return 1 if $Secrets;
    $Secrets = {};
    
    if (not isRoot()) {
	y2error("You have to have root permisions to acces to secret.tdb");
	return 0;
    }
    
    # if the secrets file does not exist at all, return empty map
    my $res = SCR->Read(".target.stat", secret_tdb);
    unless (defined $res and keys %$res) {
	y2warning(secret_tdb." not found");
	return 0;
    }
    
    $res = SCR->Execute(".target.bash_output", tdbdump . " " . secret_tdb);
    if ($res->{exit}) {
	y2error("Cannot read TDB dump: " . ($res->{stdout}||"<undef>"));
	return 0;
    }
    
    my $current_key = undef;
    my @lines = split(/\n/, $res->{"stdout"});
    while (my $line = shift @lines) {
	if ($line =~ /^key = \"([^\"]*)\"$/) {
	    $current_key = $1;
	} elsif ($line =~ /^data = \"([^\"]*)\"$/) {
	    my $current_data = $1;
	    if (not defined $current_key) {
		y2error("Broken TDB dump - data without key");
		return 0;
	    }
	    $Secrets->{$current_key} = $current_data;
	    $current_key = undef;
	}
    }
    y2debug("Readed " . scalar(keys %$Secrets) . " secret keys");
    return 1;
}

# get value from samba secret db
BEGIN{$TYPEINFO{GetKey}=["function","string","string"]}
sub GetKey {
    my ($self, $key) = @_;
    
    # initialize if needed
    Read(0) or return undef unless $Secrets;

    return $Secrets->{$key};
}

# get LDAP bind passford for given admin dn
BEGIN{$TYPEINFO{GetLDAPBindPw}=["function","string","string"]}
sub GetLDAPBindPw {
    my ($self, $admin_dn) = @_;
    
    Read(0) or return undef unless $Secrets;
    my $passwd = $Secrets->{"SECRETS/LDAP_BIND_PW/$admin_dn"};
    $passwd =~ s/\00// if defined $passwd;
    unless (defined $passwd) {
	y2warning("Cannot find LDAP bind password");
    }
    return $passwd;
}

# set LDAP bind dn
# NOTE: you have to write "ldap admin dn" to smb.conf first!
BEGIN{$TYPEINFO{WriteLDAPBindPw}=["function","boolean","string"]}
sub WriteLDAPBindPw {
    my ($self, $passwd) = @_;
    
    # change password
    my $cmd = "smbpasswd -w '$passwd'";
    my $result = SCR->Execute(".target.bash_output", $cmd);
    if ($result->{exit}) {
	y2error($result->{stderr} || "Cannot set the LDAP password.");
	return 0;
    }
    $Secrets = undef;
    return 1;
}

# return trused domains / passwords pairs
BEGIN{$TYPEINFO{GetTrustedDomains}=["function",["list","string"]]}
sub GetTrustedDomains {
    my ($self) = @_;
    
    # initialize if needed
    Read(0) or return undef unless $Secrets;

    my $res = [];
    while(my ($key, $data) = each %$Secrets) {
	push @$res, $1 if $key =~ m|SECRETS/\$DOMTRUST.ACC/(.+)|;
    };
    return $res;
}

8;

# EOF
