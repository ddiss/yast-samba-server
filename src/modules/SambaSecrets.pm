#! /usr/bin/perl
# ------------------------------------------------------------------------------
# Copyright (c) 2006-2012 Novell, Inc. All Rights Reserved.
#
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of version 2 of the GNU General Public License as published by the
# Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail, you may find
# current contact information at www.novell.com.
# ------------------------------------------------------------------------------

# File:		modules/SambaSecrets.pm
# Package:	Samba server
# Summary:	Reading of /etc/samba/secrets.tdb
# Authors:	Stanislav Visnovsky <visnov@suse.cz>
#
# $Id$
#

package SambaSecrets;

use strict;
use Data::Dumper;

use YaST::YCP qw(:DATA :LOGGING);
use YaPI;

textdomain "samba-server";
our %TYPEINFO;

## Global imports
BEGIN{
YaST::YCP::Import("SCR");
}

use constant {
    secret_tdb => "/etc/samba/secrets.tdb",
    tdbdump => "/usr/bin/tdbdump",
};

our $Secrets;

# return true if effective user id is 0 (root user)
##sub isRoot {
##    return 1; #TODO
##}

# Read the current contents of the secrets file
BEGIN{$TYPEINFO{Read}=["function", "boolean"]}
sub Read {
    my ($self) = shift;
    
    return 1 if $Secrets;
    $Secrets = undef;
    
##    if (not isRoot()) {
##	y2error("You have to have root permisions to acces to secret.tdb");
##	return undef;
##
##    }
    # if the secrets file does not exist at all, return empty map
    my $res = SCR->Read(".target.stat", secret_tdb);
    unless (defined $res and keys %$res) {
	y2warning(secret_tdb." not found");
	return undef;
    }
    
    $res = SCR->Execute(".target.bash_output", tdbdump . " " . secret_tdb);
    if (!defined $res or $res->{exit}) {
	y2error("Cannot read TDB dump: " . Data::Dumper->Dump([$res], ["result"]));
	return undef;
    }
    
    my $current_key = undef;
    my @lines = split(/\n/, $res->{"stdout"});
    while (my $line = shift @lines) {
	if ($line =~ /^key = \"([^\"]*)\"$/) {
	    $current_key = $1;
	} elsif ($line =~ /^data = \"([^\"]*)\"$/) {
	    my $current_data = $1;
	    if (not defined $current_key) {
		y2error("Broken TDB dump - data without key");
		$Secrets = undef;
		return undef;
	    }
	    $Secrets->{$current_key} = $current_data;
	    $current_key = undef;
	}
    }
    y2debug("Readed " . scalar(keys %$Secrets) . " secret keys");
    return 1;
}

# get value from samba secret db
BEGIN{$TYPEINFO{GetKey}=["function","string","string"]}
sub GetKey {
    my ($self, $key) = @_;
    
    # initialize if needed
    return undef unless $Secrets or Read(0);

    return $Secrets->{$key};
}

# get LDAP bind passford for given admin dn
BEGIN{$TYPEINFO{GetLDAPBindPw}=["function","string","string"]}
sub GetLDAPBindPw {
    my ($self, $admin_dn) = @_;
    
    return undef unless $Secrets or Read(0);
    my $passwd = $Secrets->{"SECRETS/LDAP_BIND_PW/$admin_dn"};
    $passwd =~ s/\\00$// if defined $passwd;
    unless (defined $passwd) {
	y2warning("Cannot get LDAP bind password");
    }
    return $passwd;
}

# set LDAP bind dn
# NOTE: you have to write "ldap admin dn" to smb.conf first!
BEGIN{$TYPEINFO{WriteLDAPBindPw}=["function","boolean","string"]}
sub WriteLDAPBindPw {
    my ($self, $passwd) = @_;
    
    # change password
    my $cmd = "smbpasswd -w '".($passwd||"")."'";
    my $result = SCR->Execute(".target.bash_output", $cmd);
    if (!defined $result || $result->{exit}) {
	y2error("Cannot set the LDAP password: ". Data::Dumper->Dump([$result],["result"]));
	return undef;
    }
    $Secrets = undef;
    return 1;
}

# return trused domains / passwords pairs
BEGIN{$TYPEINFO{GetTrustedDomains}=["function",["list","string"]]}
sub GetTrustedDomains {
    my ($self) = @_;
    
    # initialize if needed
    return undef unless $Secrets or Read(0);

    my $res = [];
    while(my ($key, $data) = each %$Secrets) {
	push @$res, $1 if $key =~ m|SECRETS/\$DOMTRUST.ACC/(.+)|;
    };
    return $res;
}

8;

# EOF
