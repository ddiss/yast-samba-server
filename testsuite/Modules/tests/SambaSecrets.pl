#!/usr/bin/perl

use SambaSecrets;
use Data::Dumper;

## fake modules
#sub SCR::Read { shift; print "SCR::Read(",join(", ", @_),")\n"; return $read;}
#sub SCR::Execute { shift; print "SCR::Execute(",join(", ", @_),")\n";return $exec;}
sub SCR::Read { return $read;}
sub SCR::Execute { return $exec;}

$secret="{\nkey = \"x\"\ndata = \"ok\"\n}\n{\nkey = \"SECRETS/\$DOMTRUST.ACC/BAMBUS\"\ndata = \"\\01\\01\\00\\00\\00\"\n}\n{\nkey = \"SECRETS/LDAP_BIND_PW/abc\"\ndata = \"ok\\00\"\n}\n";
$secret_corrupted="{\ndata = \"\\01\"\n}\n";

## Read()
$read = undef;
print Dumper(!SambaSecrets->Read());

$read = {};
print Dumper(!SambaSecrets->Read());

$read = {size=>8};
$exec = undef;
print Dumper(!SambaSecrets->Read());

$read = {size=>8};
$exec = {exit => 256};
print Dumper(!SambaSecrets->Read());

$read = {size=>8};
$exec = {exit => 0, stdout=>$secret_corrupted};
print Dumper(!SambaSecrets->Read());

$read = {size=>8};
$exec = {exit => 0, stdout=>$secret};
print Dumper(SambaSecrets->Read());

$read = {size=>8};
$exec = {exit => 0, stdout=>$secret};
print Dumper(SambaSecrets->Read());

print Dumper($SambaSecrets::Secrets);

## GetKey()
$SambaSecrets::Secrets = undef;
$read = undef;
print Dumper(!SambaSecrets->GetKey("x"));

$SambaSecrets::Secrets = undef;
$read = {size=>8};
$exec = {exit => 0, stdout=>$secret};
print Dumper(SambaSecrets->GetKey("x"));
print Dumper(SambaSecrets->GetKey("x"));


## GetLDAPBindPw()
$SambaSecrets::Secrets = undef;
$read = undef;
print Dumper(!SambaSecrets->GetLDAPBindPw("abc"));

$SambaSecrets::Secrets = undef;
$read = {size=>8};
$exec = {exit => 0, stdout=>$secret};
print Dumper(SambaSecrets->GetLDAPBindPw("abc"));

print Dumper(!SambaSecrets->GetLDAPBindPw("xyz"));

## GetTrustedDomains
$SambaSecrets::Secrets = undef;
$read = undef;
print Dumper(!SambaSecrets->GetTrustedDomains());

$SambaSecrets::Secrets = undef;
$read = {size=>8};
$exec = {exit => 0, stdout=>$secret};
print Dumper(SambaSecrets->GetTrustedDomains());
print Dumper(SambaSecrets->GetTrustedDomains());


## WriteLDAPBindPw()
$exec = undef;
print Dumper(!SambaSecrets->WriteLDAPBindPw());

$exec = {exit => 1, stderr => "error"};
print Dumper(!SambaSecrets->WriteLDAPBindPw("secret"));

$exec = {exit => 0};
print Dumper(SambaSecrets->WriteLDAPBindPw("secret"));

