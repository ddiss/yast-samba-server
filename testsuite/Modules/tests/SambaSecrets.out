$VAR1 = 1;
$VAR1 = 1;
$VAR1 = 1;
$VAR1 = 1;
$VAR1 = 1;
$VAR1 = 1;
$VAR1 = 1;
$VAR1 = {
          'SECRETS/$DOMTRUST.ACC/BAMBUS' => '\\01\\01\\00\\00\\00',
          'x' => 'ok',
          'SECRETS/LDAP_BIND_PW/abc' => 'ok\\00'
        };
$VAR1 = 1;
$VAR1 = 'ok';
$VAR1 = 'ok';
$VAR1 = 1;
$VAR1 = 'ok';
$VAR1 = 1;
$VAR1 = 1;
$VAR1 = [
          'BAMBUS'
        ];
$VAR1 = [
          'BAMBUS'
        ];
$VAR1 = 1;
$VAR1 = 1;
$VAR1 = 1;
